Fix datadir handling

hebasto opened the pull request:
Fix #15240, see: https://github.com/bitcoin/bitcoin/issues/15240#issuecomment-487353760
Fix #15745
Fix broken `feature_config_args.py` tests (disabled by MarcoFalke@fabe28a0cdcfa13e0e595a0905e3642a960d3077). All test are enabled now.
This PR is alternative to #13621.

User's `$HOME` directory is not touched unnecessarily now.

~To make reviewing easier only `bitcoind` code is modified (neither `bitcoin-cli` nor `bitcoin-qt`).~

Refs:
- https://github.com/bitcoin/bitcoin/issues/15745#issuecomment-479852569 by **laanwj**
- #16220

promag commented on src/util/system.cpp:
```diff
@@ -784,7 +789,8 @@ void ClearDatadirCache()
 {
-    return AbsPathForConfigVal(fs::path(confPath), false);
+    fs::path path = fs::path(confPath);
+    return path.is_absolute() ? path : AbsPathForConfigVal(path, false);
```
Is this necessary? Looks like `fs::absolute` already does this in `AbsPathForConfigVal`.

hebasto replied:
Yes, it is needed to not touch default datadir unnecessarily, before all configs are read.

ryanofsky replied:
It seems like a better fix (simpler and more general) would be to add:

```c++
if (path.is_absolute()) return path;
```

as the first line of AbsPathForConfigVal.

promag commented on src/util/system.cpp:
```diff
@@ -773,6 +773,11 @@ const fs::path &GetDataDir(bool fNetSpecific)
 
+bool CheckDataDirOption()
+{
+    return gArgs.IsArgSet("-datadir") ? fs::is_directory(fs::system_complete(gArgs.GetArg("-datadir", ""))) : true;
```
This doesn't check the default data dir.

hebasto replied:
You are right. This behavior is international.

ryanofsky replied:
In commit "Add CheckDataDirOption() function" (a45f692366d7bb4814af83e3b645c9e5a5a734e6)

I don't think `IsArgSet` is the right function to call here. If somebody has a `datadir` option specified in the config file, but wants to unset it on the command line by passing `-nodatadir` or `-datadir=""`, it seems like this should be allowed. Would suggest:

```
std::string datadir = gArgs.GetArg("-datadir", "");
return datadir.empty() || fs::is_directory(fs::system_complete(datadir));
```

And a corresponding change in GetDataDir()

hebasto replied:
> I don't think IsArgSet is the right function to call here. If somebody has a datadir option specified in the config file, but wants to unset it on the command line by passing -nodatadir or -datadir="", it seems like this should be allowed.

Good catch! Agree.

hebasto commented:
@promag Thank you for your review.
To be clear, the premature address to the default datadir, until possible `-datadir` option is read from the config file(s), is the root of all issues fixed by this PR.

promag commented:
On my system:
```
# create a file in the default data dir
touch "/Users/joao/Library/Application Support/Bitcoin"

# launch with default data dir
src/bitcoind


************************
EXCEPTION: N5boost10filesystem16filesystem_errorE
boost::filesystem::create_directory: File exists: "/Users/joao/Library/Application Support/Bitcoin"
bitcoin in AppInit()

Assertion failed: (globalChainBaseParams), function BaseParams, file chainparamsbase.cpp, line 30.
[1]    53221 abort      src/bitcoind
```

hebasto commented:
@promag 
> On my system:
> 
> ```
> # create a file in the default data dir
> touch "/Users/joao/Library/Application Support/Bitcoin"
> 
> # launch with default data dir
> src/bitcoind
> 
> 
> ************************
> EXCEPTION: N5boost10filesystem16filesystem_errorE
> boost::filesystem::create_directory: File exists: "/Users/joao/Library/Application Support/Bitcoin"
> bitcoin in AppInit()
> 
> Assertion failed: (globalChainBaseParams), function BaseParams, file chainparamsbase.cpp, line 30.
> [1]    53221 abort      src/bitcoind
> ```

The same behavior is observed on master as well. This PR does not change it.

ryanofsky reviewed (commented):
This comment doesn't seem to be true: https://github.com/bitcoin/bitcoin/pull/15864#issue-272240173

> To make reviewing easier only bitcoind code is modified (neither bitcoin-cli nor bitcoin-qt).

Since GetConfigFile is called all three places. It would also seem worrying if this were true, since it would mean the different binaries wouldn't behave consistently.

hebasto committed:
Return absolute path early in AbsPathForConfigVal

This prevents premature GetDataDir() calls, e.g., when config file is
not read yet.

hebasto commented:
@ryanofsky 
Thank you for your review. Your comment has been addressed.
Also commits for `bitcoin-qt` and tools (`bitcoin-cli`, `bitcoin-wallet`) have been added.

Would you mind re-reviewing?

ryanofsky commented on src/bitcoind.cpp:
```diff
@@ -95,8 +95,7 @@ static bool AppInit(int argc, char* argv[])
     {
-        if (!fs::is_directory(GetDataDir(false)))
-        {
+        if (!CheckDataDirOption()) {
```
In commit "Fix datadir handling in bitcoind" (160470f76f22edd1f9ef6448584381fc5626b9c4)

I can't figure out what this commit is doing. Is it fixing #15745 or #15864? Is the idea to prevent the datadir from being cached here? Can you write in the commit description what problem this is fixing and how it works?

hebasto replied:
Please see https://github.com/bitcoin/bitcoin/pull/15864#issuecomment-490952565

ryanofsky reviewed (approved):
utACK c4ab7420ef5dc4e6925519f8571a936f38720862. It's not really clear to me which bugs are being fixed by these changes, and how exactly the fixes work, but the basic idea seems to be to avoid calling `GetDataDir()` various places early on startup when it isn't necessary, to avoid caching bad values.

The changes seem worth making as optimizations or for clarity regardless of the bugs. But it'd also make sense as followup to look into:

- Whether caching datadir is a useful optimization
- Whether it makes sense to interpret a non-absolute "-conf" argument relative to the datadir

Both of these seem questionable to begin with.

hebasto commented:
@ryanofsky 
> It's not really clear to me which bugs are being fixed by these changes, and how exactly the fixes work, but the basic idea seems to be to avoid calling GetDataDir() various places early on startup when it isn't necessary, to avoid caching bad values.

The root of the reported bugs (#15240, #15745) is `GetDataDir()` tries to _create_ data directory. This step is premature when config file is not read yet.

hebasto commented on src/util/system.cpp:
```diff
@@ -1226,6 +1226,9 @@ int64_t GetStartupTime()
 
 fs::path AbsPathForConfigVal(const fs::path& path, bool net_specific)
 {
+    if (path.is_absolute()) {
```
Fixed.

promag commented:
I think we could replace the directory creation with an assertion (that datadir already exists) and create the directory right before the first call to `GetDataDir()`.

hebasto commented:
@promag 
> I think we could replace the directory creation with an assertion (that datadir already exists) and create the directory right before the first call to `GetDataDir()` - probably on init.

Is it a bit out of the scope of this PR?

hebasto commented:
@ryanofsky 
All your comments have been addressed.

ryanofsky commented on src/util/system.cpp:
```diff
@@ -753,8 +753,9 @@ const fs::path &GetDataDir(bool fNetSpecific)
 
-    if (gArgs.IsArgSet("-datadir")) {
-        path = fs::system_complete(gArgs.GetArg("-datadir", ""));
+    std::string datadir = gArgs.GetArg("-datadir", "");
```
In commit "Add CheckDataDirOption() function" (5df1581b9ee6f13593b33fa9629ea9fbadcad732)

Ideally commit message might also mention that it changes the behavior of the GetDataDir function. The effect I think is to now allow `-nodatadir` or `-datadir=''` options on the command line to override any datadir specified in the config file and reset it to default. In theory this could even be mentioned in release notes, but it's probably too niche.

hebasto replied:
> The effect I think is to now allow `-nodatadir` or `-datadir=''` options on the command line to override any datadir specified in the config file and reset it to default.

Unfortunately, neither `-nodatadir` nor `-datadir=''` on the command line can unset datadir specified in the config file or reset it to default. See: #16097.

MarcoFalke replied:
Then, why is this changed?

hebasto replied:
The [initial idea](https://github.com/bitcoin/bitcoin/pull/15864#discussion_r281785729) was:
> I don't think `IsArgSet` is the right function to call here.

If we recall:
https://github.com/bitcoin/bitcoin/blob/8c69fae94410f54bad13be0f34d54370fddbf4b3/src/util/system.cpp#L470-L474

using `IsArgNegated` is (logically) inappropriate for `-datadir` option, IMO.

Also (but minor), the suggested code uses only one function call (`GetArg`) instead of two calls (`IsArgSet` and `GetArg`).

hebasto replied:
@ryanofsky @MarcoFalke 
> The effect I think is to now allow `-nodatadir` or `-datadir=''` options on the command line to override any datadir specified in the config file and reset it to default.

Actually, `-datadir=''` (with empty string) or `-datadir` will reset it to default. So, going to close #16416 in favor of this PR.

ryanofsky reviewed (approved):
utACK 9bddf12763e1bcc41c176f8b7f0b883efed49823. Changes since last review are adding commit descriptions, documenting new `CheckDataDirOption()` method, and handling `-nodatadir` and `-datadir=""` better.

This PR should be tagged as needing release notes if release notes aren't added before it is merged in a `doc/release-notes-15864.md` file. I think you could take the description from https://github.com/bitcoin/bitcoin/pull/15864#issuecomment-490952565 and basically turn it into a release note, describing the class of bugs that are fixed and not mentioning specific function names.

hebasto commented:
@MarcoFalke would you mind reviewing this PR?

hebasto commented:
Rebased after #16205 merged.

MarcoFalke commented:
Concept ACK

ryanofsky reviewed (approved):
utACK 4f0de6214ae591bfb37ef745140f45fcc97aec33. No changes since last review except rebase and replacing fprintf with tfm::format.

hebasto commented:
Rebased on top of #16300.

MarcoFalke commented on src/util/system.cpp:
```diff
@@ -1199,6 +1199,9 @@ int64_t GetStartupTime()
 {
+    if (path.is_absolute()) {
+        return path;
+    }
```
Would be nice if each commit had a test that failed before compilation and passes after. That might simplify review, no?

hebasto replied:
Do tests enabled in 5124b1cef24a9b30876ddedf922dfdb555bd4ace do this job?

MarcoFalke replied:
Can they be enabled in this commit (the first commit)?

hebasto replied:
Do you mean to combine the "Enable all tests in feature_config_args.py " commit (5124b1cef24a9b30876ddedf922dfdb555bd4ace) into the "Return absolute path early in AbsPathForConfigVal" commit (c21150fd6f598196fae93cf4bddc6e7625a43660)?

Master branch plus this combined commit will fail the test.
And the test will be passed on top of the 584621f9a81b6cc547c8f1365f1b358866e9a335 and b921f6ee951f0ad6d87eeeb372136a7314af0b0e commits.

Did I understand your intention correctly?

ryanofsky reviewed (approved):
utACK 5f55360ea36487131a14dbc0a2518ec3babc1580. No changes since last review other than rebase and comment tweak.

hebasto commented:
Rebased.

hebasto commented:
@laanwj Would you mind reviewing this PR?

ryanofsky reviewed (approved):
utACK fd7310f24422901e0d4cf82c93f04b2b1b458865. No changes since last review other than rebase after #16291

MarcoFalke commented:
The 09a307183aca7b2f6e1a14caf4430a1fc546e704 commit would have to be in the test commit or before it, otherwise the test running with the gui might fail on the test commit?

hebasto committed:
Add CheckDataDirOption() function

hebasto committed:
Fix datadir handling in bitcoind

This prevents premature tries to access or create the default datadir.
This is useful when the -datadir option is specified and the default
datadir is unreachable.

hebasto committed:
Fix datadir handling in bitcoin-qt

This prevents premature tries to access or create the default datadir.
This is useful when the -datadir option is specified and the default
datadir is unreachable.

hebasto committed:
Fix datadir handling in bitcoin-cli

This prevents premature tries to access or create the default datadir.
This is useful when the -datadir option is specified and the default
datadir is unreachable.

hebasto committed:
Use CheckDataDirOption() for code uniformity

All other clients and tools use CheckDataDirOption() rather
fs::is_directory(GetDataDir(false)) for the first datadir check.

hebasto commented:
Rebased.

@MarcoFalke 
> The 09a3071 commit would have to be in the test commit or before it, otherwise the test running with the gui might fail on the test commit?

Commits have been rearranged. The test commit is the last one now.

hebasto committed:
Enable all tests in feature_config_args.py

hebasto commented:
@MarcoFalke https://github.com/bitcoin/bitcoin/issues/15240#issuecomment-541334088

Label "Needs backport 0.18" ?
